src/main/java/org/anarres/parallelgzip/ParallelGZIPOutputStream.java

```diff
@@ -44,25 +37,35 @@ private static DeflaterOutputStream newDeflaterOutputStream(@Nonnull OutputStrea
         return new DeflaterOutputStream(out, deflater, 512, true);
     }
 
-    private static class Block implements Callable<byte[]> {
+    private static class State {
+        private final Deflater def = newDeflater();
+        private final ByteArrayOutputStreamExposed buf = new ByteArrayOutputStreamExposed(SIZE);
+        private final DeflaterOutputStream str = newDeflaterOutputStream(buf, def);
+    }
+
+    /** This ThreadLocal avoids the recycling of a lot of memory, causing lumpy performance. */
+    private static final ThreadLocal<State> STATE = new ThreadLocal<State>() {
+        @Override
+        protected State initialValue() {
+            return new State();
+        }
+    };
 
-        private static class State {
+    /* Allow write into byte[] directly */
+    private static class ByteArrayOutputStreamExposed extends ByteArrayOutputStream {
+        private ByteArrayOutputStreamExposed(int size) {
+            super(size);
+        }
 
-            private final Deflater def = newDeflater();
-            private final ByteArrayOutputStream buf = new ByteArrayOutputStream(SIZE);
-            private final DeflaterOutputStream str = newDeflaterOutputStream(buf, def);
+        private void writeTo(byte[] buf) {
+            System.arraycopy(this.buf, 0, buf, 0, count);
         }
-        /** This ThreadLocal avoids the recycling of a lot of memory, causing lumpy performance. */
-        private static final ThreadLocal<State> STATE = new ThreadLocal<State>() {
-            @Override
-            protected State initialValue() {
-                return new State();
-            }
-        };
-        public static final int SIZE = 64 * 1024;
+    }
+
+    private static class Block implements Callable<Block> {
         // private final int index;
-        private final byte[] in = new byte[SIZE];
-        private int in_length = 0;
+        private final byte[] buf = new byte[SIZE * 2];
```

Comment:
This is one potential option for static allocation; I don't know that it guarantees success, though. Do we have any sort of hard guarantee that for a datum of size N, the compressed size will be less than N * k for any k?

What did you think of the memory-allocation-improvements branch I made? I took your code from #5, applied most of my review comments, and left only this. I'm tempted to go from there, and make this a dynamic allocator, starting at say 1.1.

My suspicion remains that it's not the GC cost, as eden space is essentially free, but it's the array initialization cost of this array, which is making the process slow. Either way, the fundamentals of your fix apply.